unitary/examples/quantum_chinese_chess/move.py

```diff
@@ -217,3 +220,418 @@ def effect(self, *objects) -> Iterator[cirq.Operation]:
+                # could fire only if the quantum path piece is empty.
+                if not world.pop(quantum_path_pieces_0)[0].value:
+                    # Reset it to be classically empty.
+                    quantum_path_pieces_0[0].reset()
+                    could_capture = True
+            else:
+                # We add a new ancilla to indicate whether the capture could happen (value 1 means it could).
+                capture_ancilla = world._add_ancilla(f"{source_0.name}{target_0.name}")
+                control_objects = [source_0] + quantum_path_pieces_0
+                conditions = [1] + [0] * len(quantum_path_pieces_0)
+                # We flip the ancilla only if the source is there and all quantum path pieces are empty.
+                alpha.quantum_if(*control_objects).equals(*conditions).apply(
+                    alpha.Flip()
+                )(capture_ancilla)
+                # We measure this ancilla to determine if the cannon fire could be made.
+                could_capture = world.pop([capture_ancilla])[0]
+            if not could_capture:
+                print("Cannon fire not applied: the path turns out to be blocked.")
+                return iter(())
+            # Apply the capture.
+            # Reset the target qubit.
+            world.unhook(target_0)
+            # Reset the classical properties of the target.
+            target_0.reset()
+            alpha.PhasedMove()(source_0, target_0)
+            # Move the classical properties of the source piece to the target piece.
+            target_0.reset(source_0)
+            source_0.reset()
+            # Force measure all quantum_path_pieces_0 to be empty.
+            for path_piece in quantum_path_pieces_0:
+                if path_piece.is_entangled:
+                    # We check if the piece is entangled since in the case len(quantum_path_pieces_0) == 1
+                    # the force_measurement has already been made.
+                    world.force_measurement(path_piece, 0)
+                    path_piece.reset()
+            return iter(())
+        else:
+            # In the case where there are no classical path piece but only quantum
+            # path piece(s), the cannon could fire and capture only if there is exactly
+            # one quantum path piece being occupied.
+            could_capture = False
+            # We loop over all quantum path pieces and check if it could be the only
+            # occupied piece. The fire could be made if it does, otherwise not.
+            # TODO(): think a more efficient way of implementing this case.
+            for index, expect_occupied_path_piece in enumerate(quantum_path_pieces_0):
+                # TODO(): consider specific cases to save the ancilla.
+                # Add a new ancilla to indicate whether the fire could be made (value = 1 means it could).
+                capture_ancilla = world._add_ancilla(
+                    f"{expect_occupied_path_piece.name}"
```

Comment:
is `expect_occupied_path_piece.name` not guaranteed to be a string?

Reply:
Yes it is a string. It's fixed now. Thanks.